nmostesting/GenericTest.py

```diff
@@ -383,6 +397,97 @@ def validate_schema(self, payload, schema):
     def do_request(self, method, url, **kwargs):
         return TestHelper.do_request(method=method, url=url, **kwargs)
 
+    def make_key_cert_files(self, cert_file, key_file):
+        # create a key pair
+        k = crypto.PKey()
+        k.generate_key(crypto.TYPE_RSA, 2048)
+
+        # create cert
+        cert = crypto.X509()
+        cert.set_version(2)
+        cert.get_subject().C = "GB"
+        cert.get_subject().ST = "England"
+        cert.get_subject().O = "NMOS Testing Ltd"  # noqa: E741
+        ca_cert_subject = cert.get_subject()
+        ca_cert_subject.CN = "ca.testsuite.nmos.tv"
+        cert.set_issuer(ca_cert_subject)
+        cert.get_subject().CN = "mocks.testsuite.nmos.tv"
+        cert.set_serial_number(x509.random_serial_number())
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(10*365*24*60*60)
+        cert.set_pubkey(k)
+        # get Root CA key
+        capkey = open(CONFIG.KEY_TRUST_ROOT_CA, "r").read()
+        ca_pkey = crypto.load_privatekey(crypto.FILETYPE_PEM, capkey)
+        # get Root CA cert
+        cacert = open(CONFIG.CERT_TRUST_ROOT_CA, "r").read()
+        ca_cert = crypto.load_certificate(crypto.FILETYPE_PEM, cacert)
+        # create cert extension
+        san = ["DNS:mocks.{}".format(CONFIG.DNS_DOMAIN), "DNS: nmos-mocks.local"]
+        cert_ext = []
+        cert_ext.append(crypto.X509Extension(b'subjectKeyIdentifier', False, b'hash', cert))
+        cert_ext.append(crypto.X509Extension(b'authorityKeyIdentifier', False, b'keyid,issuer:always', issuer=ca_cert))
+        cert_ext.append(crypto.X509Extension(b'basicConstraints', False, b'CA:FALSE'))
+        cert_ext.append(crypto.X509Extension(b'keyUsage', True, b'digitalSignature, keyEncipherment'))
+        cert_ext.append(crypto.X509Extension(b'subjectAltName', False, ','.join(san).encode()))
+        cert.add_extensions(cert_ext)
+
+        # sign cert with Intermediate CA key
+        cert.sign(ca_pkey, 'sha256')
+
+        # write chain certificate file
+        if cert_file is not None:
+            with open(cert_file, "wt") as f:
+                f.write(crypto.dump_certificate(crypto.FILETYPE_PEM, cert).decode("utf-8"))
+                f.write(cacert)
+        # write private key file
+        if key_file is not None:
+            with open(key_file, "wb") as f:
+                pem = k.to_cryptography_key().private_bytes(
+                    encoding=serialization.Encoding.PEM,
+                    format=serialization.PrivateFormat.TraditionalOpenSSL,
+                    encryption_algorithm=serialization.NoEncryption()
+                )
+                f.write(pem)
+
+    def start_secondary_authorization_server(self, auth):
```

Comment:
For IS-04-01, where we need multiple Registration API instances, I think we start all of them up front, but the ones we're not interested in are just not advertised via DNS-SD and set to return `5xx` from their API endpoints.

What are the pros and cons of doing that vs. this?

Reply:
Cannot use the same technique as the way how multi-registries was done. It is because the new instance of the Authorization server is required to use a new set of private key and certificate.  This new private key is used to generate the new token, and the node will not be able to decode. That will trigger the node to fetch the new public key from the auth server for the test.  I don't know how I can change the key and cert dynamically that is why I choose to restart the auth server.

Reply:
Why can't primary and secondary be started with their own keys/certs at the beginning?

Reply:
The `do_test_no_matching_public_key_authorization` test requires the secondary auth server to use a different private key/cert every time to run the test.

Reply:
Ah, makes sense! Thanks